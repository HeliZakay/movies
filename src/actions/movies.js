import uuid from "uuid";
import database from "../firebase/firebase";
import moment from "moment";


// ADD_MOVIE
export const addMovie = (movie = {
        personName: "",
        movieName: "",
        content: "",
        score: 0,
        createdAt: 0
}) => {
    return ({
      type: 'ADD_MOVIE',
      movie: {
        ...movie,
        id: uuid()
      }
    });
  };


export const startAddMovie = (movieData = {}) => {
  return (dispatch) => {
    const {
      movieName="",
      personName="",
      score=0,
      content="",
      createdAt="0"      
     } = movieData;
     const movie = {movieName, personName, score, content, createdAt: moment(createdAt).format() };
     return database.ref("movies").push(movie).then( (ref) => {
        dispatch(addMovie({
          id: ref.key,
          ...movie
        }));
     });
  };
};

// EDIT_MOVIE
export const editMovie = (id, updates) => ({
  type: 'EDIT_MOVIE',
  id,
  updates
});


  // REMOVE_MOVIE
  export const removeMovie = ({ id } = {}) => ({
    type: 'REMOVE_MOVIE',
    id
  });
  