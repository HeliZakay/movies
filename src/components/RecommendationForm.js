import React from "react";
import moment from "moment";

export default class RecommendationForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            movieName: props.movie ? props.movie.movieName : "",
            personName:props.movie ? props.movie.personName : "",
            content: props.movie? props.movie.content: "",
            score: props.movie? props.movie.score: 7,
            error:"",
            createdAt: props.movie? props.movie.createdAt: moment(),
            // watchList: props.movie? props.movie.watchList: false
        }
    }
    onMovieNameChange =(event) => {
        const movieName = event.target.value;
        this.setState( () => ({movieName}));
    };
    onPersonNameChange =(event) => {
        const personName = event.target.value;
        this.setState( () => ({personName}));
    };
    onContentChange = (event) => {
        const content = event.target.value;
        this.setState( () => ({content}));
    }
    onScoreChange = (event) => {
        const score = event.target.value;
        this.setState( () => ({score}));
    }
    onSubmit = (event) => {
        event.preventDefault();
        if (!this.state.movieName || !this.state.score) {
            this.setState( () => ({error: "Please fill in movie name and a movie rating" }));
        } else {
            this.setState( () => ({error: ""}));
            this.props.onSubmit({
                movieName: this.state.movieName,
                score: parseInt(this.state.score),
                createdAt: this.state.createdAt,
                personName: this.state.personName,
                content: this.state.content,
                // watchList: this.state.watchList
            });
        }
    };

    render() {
        return (
            <div>
                {this.state.error && <p><em>{this.state.error}</em></p>}
                <form onSubmit={this.onSubmit}>
                    <input 
                        type="text"
                        placeholder="Movie Name"
                        autoFocus
                        value = {this.state.movieName}
                        onChange= {this.onMovieNameChange}
                    />
                    
                    <label> Movie Rating: </label>
                     <input 
                        type="number"                        
                        value={this.state.score}
                        onChange={this.onScoreChange}
                        min="1"
                        max="10"
                    />
                    <br />
                    <br />
                    <textarea
                        placeholder="Write your review for the movie (optional)"
                        value={this.state.content}
                        onChange={this.onContentChange}>
                    </textarea>
                    <br />
                    <br />
                    <input 
                        type="text"     
                        placeholder="Your name (optional)"                   
                        value={this.state.personName}
                        onChange={this.onPersonNameChange}
                    />
                     <br />
                    <br />
                    <button >Recommend!</button>
                </form>
            </div>
        );
    }
}