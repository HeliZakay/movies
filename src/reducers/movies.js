// Movie Reducer


export default (state = [], action) => {
    switch (action.type) {
      case 'ADD_MOVIE':
        return [
          ...state,
          action.movie
        ];
        case 'EDIT_MOVIE':
          return state.map((movie) => {
            if (movie.id === action.movieId) {
              const updatedReviewsArray =  movie.reviews.map((review) => {
                if (review.id === action.reviewId) {
                  return {
                    ...review,
                    ...action.updatedReview
                  };
                }
                else {
                  return review;
                };
              });
              return {
                ...movie,
                reviews: updatedReviewsArray
              }
            } else {
              return movie;
            };
          });
      case 'REMOVE_MOVIE':
        return state.map((movie) => {
          if (movie.id === action.movieId){
           return {
              ...movie,
              reviews: movie.reviews.filter((review) => review.id !== action.reviewId)
            };
          } else {
            return movie;
          }
        });
      case "REMOVE_MOVIE_COMPLETELY":
        return state.filter((movie) => {
          return movie.id !== action.movieId;
        });
      case "SET_MOVIES":
        return action.movies;
      case "ADD_REVIEW":
          return state.map((movie) => {
            if (movie.id === action.movieId) {
              const updatedReviewsArray = [action.review, ...movie.reviews];
                return {
                  ...movie, 
                  reviews: updatedReviewsArray
              }
            }
            else {return movie};
          });
      case "GIVE_STAR_TO_REVIEW":
        return state.map((movie) => {
          if(movie.id === action.movieId) {
            const updtaedReviews = movie.reviews.map((review) => {
              if (review.id === action.reviewId) {
                let updatedStarsArray;
                if (review.stars) {
                  updatedStarsArray = [...review.stars, action.user];
                } else {
                  updatedStarsArray = [action.user];
                }
                return {
                  ...review,
                  stars: updatedStarsArray
                }
              } else {
                return review;
              }
            });
            return {
              ...movie,
              reviews: updtaedReviews
            }
          } else {
            return movie;
          }
        })
      default:
        return state;
    }
  };
