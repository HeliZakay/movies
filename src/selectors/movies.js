import FriendDialog from "../components/FriendDialog";

  export const getHname = (movies, movieName) => {
    const movie = movies.filter((movie) => {
      return movie.movieName === movieName;
    });
    if (movie && movie[0] && movie[0].hname) {
      return movie[0].hname;
    }
    else {
      return undefined;
    }
  }
  
  
  export const didIGaveStarToReview = (stars, uid) => {
    let result = false;
    if (stars) {
      stars.forEach((star) => {
        if (star.uid === uid) {
          result = true;
        }
      });
    }
    return result;
  }
  
  export const sortReviewsByDateAndPerson = (reviews, person) => { 
    if (!person) {
      return reviews.sort((a,b) => { 
        return a.createdAt < b.createdAt? 1:-1;
      });
    } else {
      const sortedByDate = reviews.sort((a,b) => { 
        return a.createdAt < b.createdAt? 1:-1;
      });
     const personReview =  reviews.filter((review) => {
        return review.personName.toLowerCase().includes(person.toLowerCase());
      });
      return [...personReview, ...sortedByDate.filter((review)=> {
        return !review.personName.toLowerCase().includes(person.toLowerCase());
      })];
    }   
  };

  
const computeAverageScore = (movie) => {
  let sum = 0;
  movie.reviews.forEach((review) => {
      sum += Number(review.score);
  });
  const result = sum/movie.reviews.length;
  return  Math.round(result*2)/2;;
}

// Get visible movies
export default (movies, { text, sortBy, person, genres}, language) => {
    return movies.filter((movie) => {
      let textMatch;
      if (language === "English") {
        textMatch = movie.movieName.toLowerCase().includes(text.toLowerCase());
      } else {
        textMatch = String(movie.hname).includes(text);
      }
      let personMatch = false;
      movie.reviews.forEach((review) => {
        if (review.personName.toLowerCase().includes(person.toLowerCase())) {
          personMatch = true;
        }
      });
      let genreMatch = false;
      if (!genres) {
        genreMatch = true;
      } else if (genres.length ===0) {
        genreMatch = true;
      } else if (!movie.imdbData.Genre) {
        genreMatch = false;
      } else {
        const genresArray = movie.imdbData.Genre.split(", ");
        genresArray.forEach((genre) => {
          if (genres.includes(genre)) {
            genreMatch = true;
          }
        });
      }
      return genreMatch && textMatch && personMatch;
    }).map((movie) => {
      const sorted = sortReviewsByDateAndPerson(movie.reviews, person);
      return {...movie, reviews: sorted};
    }).sort((a, b) => {
      if (sortBy === 'date') {
        return a.reviews[0].createdAt < b.reviews[0].createdAt ? 1 : -1;
      } else if (sortBy === 'score') {
        return computeAverageScore(a) < computeAverageScore(b) ? 1 : -1;
      }
    });
  };

  export const getMovieById = (movies, movieId) => {
    let result = undefined;
    movies.forEach((movie) => {
      if(movie.id === movieId) {
        result = movie;
      }
    });
    return result;
  };

  export const getMovieReviews = (movieId, movies) => {
    let reviews = [];
    movies.forEach((movie) => {
      if (movie.id === movieId) {
        reviews=movie.reviews;
      }
    });
    return reviews;
  }

  export const isMovieNameExistsAlready = (movieName, movies) => {
    let result = undefined;
    movies.forEach((movie) => {
      if (movie.movieName === movieName) {
        result=movie.id;
      }
    });
    return result;
  }
export const findMovie = (movies, movieName, personName, uid, type) => {
  const movieArr =  movies.filter((movie) => {
    return movie.movieName === movieName;
  });  
  if(movieArr && movieArr[0]) {
    const movie = movieArr[0];
    const oldReviews = movie.reviews;
    const filteredReviews = oldReviews.filter((review) => {
      if(type==="newReview") {
        return review.personName !== personName;
      }
      else {
        return review.userUid !== uid;
      }
    });
    const firstReview = oldReviews.filter((review) => {
      if(type==="newReview") {
        return review.personName === personName;
      }
      else {
        return review.userUid === uid;
      }
    });
    const sortedReviews= [...firstReview, ...filteredReviews];
    return {
      ...movie,
      reviews: sortedReviews
    };
  } else {
    return undefined;
  }
  
}

export const getUnseenMovies = (movies, uid, watchlist) => {
  return movies.filter((movie)=> {
    const reviewsUids = movie.reviews.map((review)=> {
      return review.userUid;
    });
    return (!reviewsUids.includes(uid) && !watchlist.includes(movie.id) && computeAverageScore(movie) >=7);
  });
}

export const takeTop = (array, num) => {
  if (array.length <=num) {
    return array;
  }
  else {
    return array.slice(0,num);
  }
}

export const didFriendReviewedMovie = (reviews, friendId) => {
  
  const result = reviews.filter((review) => {
    return review.userUid === friendId;
  });
  return result.length > 0;

}

const didSendReviewToMovie = ({reviews, uid}) => {
  let reviewId;
  reviews.forEach((review) => {
    if (review.userUid === uid) {
      reviewId = review.id;
    }
  });
  return reviewId;
}

export const getMyMovies = (movies, uid) => {
  return movies.filter((movie) => {
    const myReviewId = didSendReviewToMovie({reviews: movie.reviews, uid});
    if (myReviewId) {
      let hightRatedByMe = false;
      movie.reviews.forEach((review) => {
        if(review.id === myReviewId && review.score >=8){
          hightRatedByMe = true;
        }
      });
      return hightRatedByMe;
    } else {
      return false;
    }
  });
} 