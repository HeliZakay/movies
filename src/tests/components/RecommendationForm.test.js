import React from "react";
import {shallow} from "enzyme";
import RecommendationForm from "../../components/RecommendationForm";
import movies from "../fixtures/movies";
import moment from "moment";

test("should render RecommendationForm correctly", () => {
    const wrapper = shallow(<RecommendationForm />);
    expect(wrapper).toMatchSnapshot();
});

test("should render RecommendationForm correctly with expense data", () => {
    const wrapper = shallow(<RecommendationForm movie={movies[1]} />);
    expect(wrapper).toMatchSnapshot();
});

test("should render error for invalid form submission", () => {
    const wrapper = shallow(<RecommendationForm />);
    expect(wrapper).toMatchSnapshot();
    wrapper.find("form").simulate("submit", {
        preventDefault: () => {}
    });
    expect(wrapper.state("error").length).toBeGreaterThan(0);
    expect(wrapper).toMatchSnapshot();
});

test("should set movieName on input change", () => {
    const value = "new movie name";
    const wrapper = shallow(<RecommendationForm />);
    wrapper.find("input").at(0).simulate("change", {
        target: { value }
    });
    expect(wrapper.state("movieName")).toBe(value);
});

test("should set content on textarea change", () => {
    const value = "new content";
    const wrapper = shallow(<RecommendationForm />);
    wrapper.find("textarea").simulate("change", {
        target: { value }
    });
    expect(wrapper.state("content")).toBe(value);
});

test("should set personName on input change", () => {
    const value = "new person name";
    const wrapper = shallow(<RecommendationForm />);
    wrapper.find("input").at(2).simulate("change", {
        target: { value }
    });
    expect(wrapper.state("personName")).toBe(value);
});


test("should set score on input change", () => {
    const value = 9;
    const wrapper = shallow(<RecommendationForm />);
    wrapper.find("input").at(1).simulate("change", {
        target: { value }
    });
    expect(wrapper.state("score")).toBe(value);
});

test("should call onSubmit prop for valid form submission", () => {
    const onSubmitSpy = jest.fn();
    const wrapper = shallow(<RecommendationForm movie={movies[0]} onSubmit={onSubmitSpy} />);
    wrapper.find("form").simulate("submit", {
        preventDefault: () => {}
    });
    expect(wrapper.state("error")).toBe("");
    expect(onSubmitSpy).toHaveBeenLastCalledWith({
        content: movies[0].content,
        movieName: movies[0].movieName,
        personName: movies[0].personName,
        score: movies[0].score,
        createdAt: movies[0].createdAt,
        watchList: movies[0].watchList
    });
});