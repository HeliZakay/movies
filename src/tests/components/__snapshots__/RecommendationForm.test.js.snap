// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render RecommendationForm correctly 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Movie Name"
      type="text"
      value=""
    />
    <label>
       Movie Rating: 
    </label>
    <input
      max="10"
      min="1"
      onChange={[Function]}
      type="number"
      value={7}
    />
    <br />
    <br />
    <textarea
      onChange={[Function]}
      placeholder="Write your review for the movie (optional)"
      value=""
    />
    <br />
    <br />
    <input
      onChange={[Function]}
      placeholder="Your name (optional)"
      type="text"
      value=""
    />
    <br />
    <br />
    <button>
      Recommend!
    </button>
  </form>
</div>
`;

exports[`should render RecommendationForm correctly with expense data 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Movie Name"
      type="text"
      value="You've got mail"
    />
    <label>
       Movie Rating: 
    </label>
    <input
      max="10"
      min="1"
      onChange={[Function]}
      type="number"
      value={10}
    />
    <br />
    <br />
    <textarea
      onChange={[Function]}
      placeholder="Write your review for the movie (optional)"
      value="the movie is OK"
    />
    <br />
    <br />
    <input
      onChange={[Function]}
      placeholder="Your name (optional)"
      type="text"
      value="Michal"
    />
    <br />
    <br />
    <button>
      Recommend!
    </button>
  </form>
</div>
`;

exports[`should render error for invalid form submission 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Movie Name"
      type="text"
      value=""
    />
    <label>
       Movie Rating: 
    </label>
    <input
      max="10"
      min="1"
      onChange={[Function]}
      type="number"
      value={7}
    />
    <br />
    <br />
    <textarea
      onChange={[Function]}
      placeholder="Write your review for the movie (optional)"
      value=""
    />
    <br />
    <br />
    <input
      onChange={[Function]}
      placeholder="Your name (optional)"
      type="text"
      value=""
    />
    <br />
    <br />
    <button>
      Recommend!
    </button>
  </form>
</div>
`;

exports[`should render error for invalid form submission 2`] = `
<div>
  <p>
    <em>
      Please fill in both movie name and a movie rating
    </em>
  </p>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Movie Name"
      type="text"
      value=""
    />
    <label>
       Movie Rating: 
    </label>
    <input
      max="10"
      min="1"
      onChange={[Function]}
      type="number"
      value={7}
    />
    <br />
    <br />
    <textarea
      onChange={[Function]}
      placeholder="Write your review for the movie (optional)"
      value=""
    />
    <br />
    <br />
    <input
      onChange={[Function]}
      placeholder="Your name (optional)"
      type="text"
      value=""
    />
    <br />
    <br />
    <button>
      Recommend!
    </button>
  </form>
</div>
`;
